demos/coreMQTT/mqtt_demo_connection_sharing.c

```diff
@@ -0,0 +1,1962 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
+ */
+
+/*
+ * Demo for showing use of the managed MQTT API shared between multiple tasks.
+ * This demo uses a thread safe queue to hold commands for interacting with the
+ * MQTT API. A command task processes commands from the queue while other tasks
+ * enqueue them. This task enters a loop, during which it processes commands from
+ * the command queue. If a termination command is received, it will break from
+ * the loop. In addition to the command task, this demo uses one task for
+ * publishing messages to the MQTT broker and another for receiving them via
+ * an MQTT subscription. The publisher task creates a series of publish operations
+ * to push to the command queue, which are then executed by the command task.
+ * The subscriber task subscribes to a topic filter matching the topics published
+ * on by the publisher, and then loops while waiting for publish messages to be
+ * received. Each task has a queue to hold received publish messages,
+ * and the command task pushes incoming publishes to the queue of each task
+ * that is subscribed to the incoming topic.
+ */
+
+/* Standard includes. */
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+#include "queue.h"
+
+/* Demo Specific configs. */
+#include "mqtt_demo_connection_sharing_config.h"
+
+/* MQTT library includes. */
+#include "core_mqtt.h"
+#include "core_mqtt_state.h"
+
+/* Retry utilities include. */
+#include "retry_utils.h"
+
+/* Transport interface implementation include header for TLS. */
+#include "transport_secure_sockets.h"
+
+/* Credentials for the MQTT broker endpoint. */
+#include "aws_clientcredential.h"
+
+/* Include header for root CA certificates. */
+#include "iot_default_root_certificates.h"
+
+/**
+ * These configuration settings are required to run the demo.
+ */
+#ifndef democonfigCLIENT_IDENTIFIER
+
+/**
+ * @brief The MQTT client identifier used in this example.  Each client identifier
+ * must be unique so edit as required to ensure no two clients connecting to the
+ * same broker use the same client identifier.
+ */
+    #define democonfigCLIENT_IDENTIFIER    clientcredentialIOT_THING_NAME
+#endif
+
+/* Provide default values for undefined configs. */
+#ifndef democonfigMQTT_BROKER_ENDPOINT
```

Comment:
This is a strange pattern to see in a demo. This is usually what a library does if you are not sure the user will define these things. 

I would suggest that all of this stuffs should be in a DemoConfig.h file or similar which is defining what you want defined.

The demo should be a "demo" of how a user is expected to use the library. I do not expect anyone writing a real application to need config #ifdefs to add in config defaults for things they did not configure. I would expect a config file somewhere, and with that config file a description of what each config item does.

Lots of examples in the FreeRTOS Kernel demos of this e.g. https://github.com/FreeRTOS/FreeRTOS/blob/master/FreeRTOS/Demo/WIN32-MSVC/main_blinky.c

Reply:
I understand. I followed the same pattern as the other demos which had comments in the [demo config files](https://github.com/aws/amazon-freertos/blob/master/vendors/pc/boards/windows/aws_demos/config_files/mqtt_demo_mutual_auth_config.h) and the ifndef in the actual [demo file](https://github.com/aws/amazon-freertos/blob/master/demos/coreMQTT/mqtt_demo_mutual_auth.c#L94). Should that be changed for all demos?